Adds thin provisioning support for new hard disks in VMware driver

marnovdm opened the pull request:
This pull request adds a configurable option 'thin_provision' to the VMware driver for Salt Cloud. I use this with a cloud profile like this:

```
myprofile:
  clonefrom: some-vm
  provider: my-cloud
  devices:
    disk:
      Hard disk 2:
        size: 50
        thin_provision: True
```

If the option is not specified it should default to False so behaviour for existing cloud maps should be the same. I'm not a regular contributor so please let me know if I need to change something :smiley_cat:

marnovdm committed:
Add new hard disks as thin provisioned

marnovdm committed:
configurable thin_provision option

marnovdm committed:
Fix copy/paste error

marnovdm committed:
Update documentation for vmware driver

cachedout commented:
@rallytime Could you please have a look here?

rallytime commented on salt/cloud/clouds/vmware.py:
```diff
@@ -230,7 +230,7 @@ def _edit_existing_hard_disk_helper(disk, size_kb):
 
 
-def _add_new_hard_disk_helper(disk_label, size_gb, unit_number, controller_key=1000):
+def _add_new_hard_disk_helper(disk_label, size_gb, unit_number, thin_provision=False, controller_key=1000):
```
Can you move the `thin_provision` option to after the `controller_key` kwarg here? We like to try and avoid API changes where possible, just in case people are using positional kwargs somewhere, we don't what things to break.

rallytime commented on doc/topics/cloud/vmware.rst:
```diff
@@ -238,6 +238,8 @@ Set up an initial profile at ``/etc/salt/cloud.profiles`` or
         size
             Enter the size of disk in GB
+        thin_provision
+            Set to True if you want the disk to be thin provisioned. Defaults to False.
```
Can you add a `versionadded` tag here? This should look something like:
```
thin_provision
    Set to True if you want the disk to be thin provisioned. Defaults to False.
    .. versionadded:: Boron
```

rallytime commented:
@marnovdm Thank you very much for this addition! I've made a couple of inline comments for some simple things to clean up. Once those are fixed up, we'll be happy to get this merged in.

marnovdm committed:
Implement suggestions from rallytime, tries to adhere to same documentation style and fixes optional thin_provision configuration

marnovdm commented:
Thanks for the feedback @rallytime, I've tried to implement all your suggestions. I also noticed the new option wasn't actually optional the way I implemented it, I've fixed that as well.

rallytime commented:
@marnovdm Ah, you're right. Thanks for catching that. We'll get this in once tests are finished.

marnovdm commented:
Hi @rallytime, I just upgraded to 2015.8.10 and noticed this wasn't in the release yet, any estimate when this will be available?

rallytime commented:
Hi @marnovdm - Since this fix was submitted against the develop branch in December as a new addition/feature to driver, it will be available in the upcoming feature release (2016.3.0), which we are hoping to finalize and release very soon.

marnovdm commented:
Thanks for the clarification @rallytime! For future pull requests, should I use another branch than develop to get it in faster or is this the intended workflow?

rallytime commented:
@marnovdm The workflow/process we follow is that bug-fixes should go into the earliest supported release branch that the bug applies to. We merge older release branches forward into newer release branches (for example, 2015.8 --> 2016.3, the 2016.3 --> develop) so fixes can cascade up through all applicable branches.

New features, as well as major refactors, on the other hand, always go into develop so they can have time to mature and harden by people trying them out and fixing things as needed.

There's some handy docs on this process [here](https://docs.saltstack.com/en/latest/topics/development/contributing.html#which-salt-branch) for more details.
